### PR TITLE
add GroupedMetadata as a base class for Interval

### PR DESCRIPTION
**Pull request by adriangb**
The main idea here is to generalize the pattern in `Interval` so that Pydantic and similar can define their `Field` as inheriting from `GroupedMetadata`, thus anything that can parse `annotated-types` will know how to unpack it (if it is not already unpacked by the PEP-646) even if it is a custom subclass in a library like `Pydantic`.

Without this, some generic `annotated-types` parser would not know how to handle `Pydantic`'s `Field` type without specific knowledge of Pydantic.

### CODE REVIEW

File: annotated_types/__init__.py [hunk 1]
```diff
@@ -129,8 +133,37 @@ class Le(BaseMetadata):
     le: SupportsLe
 
 
+class GroupedMetadata:
```
Comment (adriangb) [inline annotated_types/__init__.py:136]: Should this be

```suggestion
class GroupedMetadata(BaseMetadata):
```

?

File: annotated_types/__init__.py [hunk 2]
```diff
@@ -129,8 +133,37 @@ class Le(BaseMetadata):
...
+    both with and without unpacking:
+
+    - `Annotated[int, Field(...)]`
+    - `Annotated[int, *Field(...)]` (PEP-646)
+    """
+    def __iter__(self) -> Iterable[BaseMetadata]:
+        return ()
+
+
+
```
Comment (adriangb) [inline annotated_types/__init__.py:164]:
```suggestion
```

**Comment by Zac-HD** [comment]
Oh, very nice!  I think that all this needs is documentation in the README as well as the lovely docstring.  A few points to clarify:

- I'd be inclined to make `__iter__` an abstractmethod, so that downstream implementors actually do implement it.
- We should also specify that `__iter__` should (to the extent possible) fully describe the constraints this metadata poses in terms of standard annotated-types constraints, and *may* include other annotations.
- And then also specify that consumers of the annotated-types standard should check for GroupedMetadata *instead of* Interval etc.

After that, let's merge and release a new version!

**Comment by adriangb** [comment]
Oh, very nice!  I think that all this needs is documentation in the README as well as the lovely docstring.  A few points to clarify:

> - I'd be inclined to make `__iter__` an abstractmethod, so that downstream implementors actually do implement it.

Sounds good to me. Can just add an `@abstractmethod` without inheriting from `abc.ABC` and have type checkers enforce it, or do we need to add the metaclass? I forget.

> - We should also specify that `__iter__` should (to the extent possible) fully describe the constraints this metadata poses in terms of standard annotated-types constraints, and *may* include other annotations.
> - And then also specify that consumers of the annotated-types standard should check for GroupedMetadata *instead of* Interval etc.

I think what you are saying is that `GroupedMetadata` should not have any meaning on its own, all of its information should be contained in the `BaseMetadata`'s that get unpacked by `__iter__()`, right?

What do you think of having `GroupedMetadata` _not_ inherit from `BaseMetadata`? I think it makes it clearer that `GroupedMetadata` in and of itself has no meaning.

**Comment by adriangb** [comment]
@Zac-HD I'm terribly sorry I hit edit instead of quote on your comment, so I edited yours instead of posting my own. I restored it now

**Comment by Zac-HD** [comment]
> Sounds good to me. Can just add an `@abstractmethod` without inheriting from `abc.ABC` and have type checkers enforce it, or do we need to add the metaclass? I forget.

Have to add the metaclass, which is most easily done by inheriting from `abc.ABC`.
 
> I think what you are saying is that `GroupedMetadata` should not have any meaning on its own, all of its information should be contained in the `BaseMetadata`'s that get unpacked by `__iter__()`, right?

Yep.

> What do you think of having `GroupedMetadata` _not_ inherit from `BaseMetadata`? I think it makes it clearer that `GroupedMetadata` in and of itself has no meaning.

That sounds good to me!

That said, it seems useful to have a single checkable class.  Maybe `BaseMetadata` should inherit from `GroupedMetadata`, and define `__iter__` as `yield self`?

**Comment by adriangb** [comment]
> Maybe `BaseMetadata` should inherit from `GroupedMetadata`, and define `__iter__` as `yield self`?

That sounds nice in theory, but I'm weary of creating class hierarchies if I don't really need them. It's really no more LOC or complexity for consumers to have two checks nested or at the same level (you need 2 anyway), I'll push that version in a few min

**Comment by adriangb** [comment]
Pushed. Like I said the difference is minor, there are always going to be 2 `isinstance` checks, it's just about what level they are at:

https://github.com/annotated-types/annotated-types/pull/12/files#diff-41d180f6f7233d175c2dfe36c45c4ea80eed4754fa7ed4eed46ccde1b2a778c7

File: tests/test_main.py [hunk 3]
```diff
@@ -98,11 +98,10 @@ def get_constraints(tp: type) -> Iterator[Constraint]:
...
         if isinstance(arg, (annotated_types.BaseMetadata, re.Pattern, slice)):
-            if isinstance(arg, annotated_types.Interval):
-                for case in arg:
-                    yield case
-            else:
-                yield arg
+            yield arg
+        elif isinstance(arg, annotated_types.GroupedMetadata):
+            for case in arg:
+                yield case
```
Comment (Zac-HD) [inline tests/test_main.py:104]:
```suggestion
            yield from arg
```

**Review by Zac-HD — COMMENTED** [review]
Implementation LGTM, pending documentation.

File: tests/test_main.py (hunk 3 above)
Comment (adriangb) [reply tests/test_main.py]: We'll now it's less LOC!

**Comment by adriangb** [comment]
I started writing the docs and it felt a bit awkward because we don't really have any documentation discussing `BaseMetadata` or how one is supposed to "consume" any of the metadata in the first place. I think this warrants it's own section of the docs and the current section should be moved into a subheading **Constraints** or something like that. Does this sound reasonable?

**Comment by Zac-HD** [comment]
Sounds reasonable; could be a follow-up or part of this PR; I'm happy with either.  If follow-up, I think this is ready to merge (though not release without docs).

**Comment by adriangb** [comment]
Sorry I merged right before your comment.

Agreed on avoiding ABCs, but I think we couldn't do much here. Protocol would be a no-go because Protocol + isinstance isn't great either.

We could always remove it and just manually raise a NotImplementedError or something.

**Comment by samuelcolvin** [comment]
> We could always remove it and just manually raise a NotImplementedError or something.

That was going to be my suggestion, then the page updated as the PR was merged :-).

I think this is fine for now, surely abcs will have to be fixed in cpython soon + my above comments.

**Comment by adriangb** [comment]
If you feel strongly, we can still change it!

**Comment by Zac-HD** [comment]
(looks like re snuck back in somehow)

Happy to do whatever on abc, so long as the error is raised when subclassing, not when you try to iterate.

**Comment by samuelcolvin** [comment]
> Happy to do whatever on abc, so long as the error is raised when subclassing, not when you try to iterate.

If we can do that, that's great with me. But is there a reliable (and fairly simple!) way to raise an error when defining subclass if a method hasn't been implemented?

If not, let's leave it as-is.

**Comment by adriangb** [comment]
You can implement `__init_subclass__` to hook into subclassing, but this won't do anything for linters/static analysis.

**Comment by Zac-HD** [comment]
I'm not too worried about linters if we can get a fully-reliable runtime check, and it looks like `__init_subclass__` would do that with less weird-abc-related performance concerns 👍

**Comment by adriangb** [comment]
I opened #16, we can continue that discussion there

**Pull request merged**